src/flint/fmpq.jl

```diff
@@ -1121,97 +1052,137 @@ function set!(c::QQFieldElemOrPtr, a::UInt, b::UInt = UInt(1))
+function numerator!(z::ZZRingElem, y::QQFieldElem)
+  ccall((:fmpq_numerator, libflint), Cvoid, (Ref{ZZRingElem}, Ref{QQFieldElem}), z, y)
+  return z
+end
+
+function denominator!(z::ZZRingElem, y::QQFieldElem)
+  ccall((:fmpq_denominator, libflint), Cvoid, (Ref{ZZRingElem}, Ref{QQFieldElem}), z, y)
+  return z
+end
+
+function mul!(c::QQFieldElemOrPtr, a::QQFieldElemOrPtr, b::QQFieldElemOrPtr)
   ccall((:fmpq_mul, libflint), Nothing,
         (Ref{QQFieldElem}, Ref{QQFieldElem}, Ref{QQFieldElem}), c, a, b)
   return c
 end
 
-function mul!(c::QQFieldElem, a::QQFieldElem, b::ZZRingElem)
+function mul!(c::QQFieldElemOrPtr, a::QQFieldElemOrPtr, b::ZZRingElemOrPtr)
   ccall((:fmpq_mul_fmpz, libflint), Nothing,
         (Ref{QQFieldElem}, Ref{QQFieldElem}, Ref{ZZRingElem}), c, a, b)
   return c
 end
 
-mul!(c::QQFieldElem, a::ZZRingElem, b::QQFieldElem) = mul!(c, b, a)
-
-function mul!(c::QQFieldElem, a::QQFieldElem, b::Int)
+function mul!(c::QQFieldElemOrPtr, a::QQFieldElemOrPtr, b::Int)
   ccall((:fmpq_mul_si, libflint), Nothing,
         (Ref{QQFieldElem}, Ref{QQFieldElem}, Int), c, a, b)
   return c
 end
 
-mul!(c::QQFieldElem, a::Int, b::QQFieldElem) = mul!(c, b, a)
+function mul!(c::QQFieldElemOrPtr, a::QQFieldElemOrPtr, b::UInt)
+  ccall((:fmpq_mul_ui, libflint), Nothing,
+        (Ref{QQFieldElem}, Ref{QQFieldElem}, Int), c, a, b)
+  return c
+end
 
+mul!(c::QQFieldElemOrPtr, a::Union{ZZRingElemOrPtr, Int, UInt}, b::QQFieldElemOrPtr) = mul!(c, b, a)
 
-function addmul!(c::QQFieldElem, a::QQFieldElem, b::QQFieldElem)
+function addmul!(c::QQFieldElemOrPtr, a::QQFieldElemOrPtr, b::QQFieldElemOrPtr)
   ccall((:fmpq_addmul, libflint), Nothing,
         (Ref{QQFieldElem}, Ref{QQFieldElem}, Ref{QQFieldElem}), c, a, b)
   return c
 end
 
-function add!(c::QQFieldElem, a::QQFieldElem, b::QQFieldElem)
+function add!(c::QQFieldElemOrPtr, a::QQFieldElemOrPtr, b::QQFieldElemOrPtr)
   ccall((:fmpq_add, libflint), Nothing,
         (Ref{QQFieldElem}, Ref{QQFieldElem}, Ref{QQFieldElem}), c, a, b)
   return c
 end
 
-function add!(c::QQFieldElem, a::QQFieldElem, b::ZZRingElem)
+function add!(c::QQFieldElemOrPtr, a::QQFieldElemOrPtr, b::ZZRingElemOrPtr)
   ccall((:fmpq_add_fmpz, libflint), Nothing,
         (Ref{QQFieldElem}, Ref{QQFieldElem}, Ref{ZZRingElem}), c, a, b)
   return c
 end
 
-add!(c::QQFieldElem, a::ZZRingElem, b::QQFieldElem) = add!(c, b, a)
-
-function add!(c::QQFieldElem, a::QQFieldElem, b::Int)
+function add!(c::QQFieldElemOrPtr, a::QQFieldElemOrPtr, b::Int)
   ccall((:fmpq_add_si, libflint), Nothing,
         (Ref{QQFieldElem}, Ref{QQFieldElem}, Int), c, a, b)
   return c
 end
 
-add!(c::QQFieldElem, a::Int, b::QQFieldElem) = add!(c, b, a)
+function add!(c::QQFieldElemOrPtr, a::QQFieldElemOrPtr, b::UInt)
+  ccall((:fmpq_add_ui, libflint), Nothing,
+        (Ref{QQFieldElem}, Ref{QQFieldElem}, Int), c, a, b)
+  return c
+end
+
+add!(c::QQFieldElemOrPtr, a::Union{ZZRingElemOrPtr, Int, UInt}, b::QQFieldElemOrPtr) = add!(c, b, a)
```

Comment:
With the proposed change above (doing conversion in add! instead of +), this should get changed to IntegerUnion as well, to only have to do conversion in the last argument

If we don't apply the above change, I would argue that this method doesn't have a right to stay, as we only allow restricted types in these functions.
Similarly, sub! doesn't handle the wrong argument order

Reply:
The case of `sub!` is a bit different, because there we can't just switch the order without loss -- we have to insert a `neg!`